pkg/core/mpt/batch.go

```diff
@@ -74,23 +76,31 @@ func (t *Trie) putBatchIntoBranch(curr *BranchNode, kv []keyValue) (Node, int, e
 	return t.addToBranch(curr, kv, true)
 }
 
-func (t *Trie) mergeExtension(prefix []byte, sub Node) Node {
+func (t *Trie) mergeExtension(prefix []byte, sub Node) (Node, error) {
 	switch sn := sub.(type) {
 	case *ExtensionNode:
 		t.removeRef(sn.Hash(), sn.bytes)
 		sn.key = append(prefix, sn.key...)
 		sn.invalidateCache()
 		t.addRef(sn.Hash(), sn.bytes)
-		return sn
+		return sn, nil
 	case *HashNode:
-		return sn
+		if sn.IsEmpty() {
+			return sn, nil
+		}
+
+		n, err := t.getFromStore(sn.Hash())
+		if err != nil {
```

Comment:
Can it really happen if trie is consistent? Maybe we should just panic here?

Reply:
No, it can't. I agree we can panic, but let's do it in another task. There are much more places in MPT which can benefit from this #2035 .

Reply:
OK, we need to think about it.